src/routes/getSkipSegmentsByHash.js

```diff
@@ -0,0 +1,125 @@
+        sqrtWeightsList.push(sqrtVote);
+        totalSqrtWeights += sqrtVote;
+    }
+
+    const randomNumber = Math.random();
+    let currentVoteNumber = 0;
+    for (let i = 0; i < sqrtWeightsList.length; i++) {
+        if (randomNumber > currentVoteNumber / totalSqrtWeights && randomNumber < (currentVoteNumber + sqrtWeightsList[i]) / totalSqrtWeights) {
+            return rows[i];
+        }
+        currentVoteNumber += sqrtWeightsList[i];
+    }
+}
+/**
+ * @param {string} prefix Lowercased hexadecimal hash prefix
+ * @param {string} hashedIP Custom hash of the visitor’s IP address
+ * @returns {Object.<string, Segment[]>}
+ */
+function getSkipSegmentsByHash(prefix, hashedIP) {
+    /** @type Row[] */
+    const rows = db.prepare('SELECT videoID, startTime, endTime, votes, UUID, category, shadowHidden FROM sponsorTimes WHERE votes >= -1 AND hashedVideoID LIKE ? ORDER BY videoID, startTime')
+            .all(prefix + '%');
+    /** @type {string[]} */
+    const onlyForCurrentUser = privateDB.prepare('SELECT videoID FROM sponsorTimes WHERE hashedIP = ?').all(hashedIP).map(row => row.videoID);
+    /** @type {Object.<string, Segment[][]>} */
+    const rowGroupsPerVideo = {};
+
+    let previousVideoID = null;
+    let previousEndTime = null;
+    for (const row of rows) {
+        /** @TODO check if this logic does what is expected. */
+        if (row.shadowHidden === 1 && onlyForCurrentUser.indexOf(row.videoID) === -1) {
+            // The current visitor’s IP did not submit for the current video.
+            // Do not send shadowHidden segments to them.
+            continue;
+        }
+        // Split up the rows per video and group overlapping segments together.
+        if (!(row.videoID in rowGroupsPerVideo)) {
+            rowGroupsPerVideo[row.videoID] = [];
+        }
+        if (previousVideoID === row.videoID && row.startTime <= previousEndTime) {
+            rowGroupsPerVideo[row.videoID][rowGroupsPerVideo[row.videoID].length - 1].push(row);
+            previousEndTime = Math.max(previousEndTime, row.endTime);
+        } else {
+            rowGroupsPerVideo[row.videoID].push([row]);
+            previousVideoID = row.videoID;
+            previousEndTime = row.endTime;
+        }
+    }
+
+    /** @type {Object.<string, Segment[]>} */
+    const output = {};
+    for (const videoID in rowGroupsPerVideo) {
+        const pickedVideosForVideoID = [];
+        for (const group of rowGroupsPerVideo[videoID]) {
+            pickedVideosForVideoID.push(pickWeightedRandomRow(group));
+        }
+        output[videoID] = pickedVideosForVideoID.map(row => ({ videoID: row.videoID, segment: [row.startTime, row.endTime], category: row.category, UUID: row.UUID }));
+    }
+    return output;
+}
+
+const minimumPrefix = config.minimumPrefix || '3';
+const maximumPrefix = config.maximumPrefix || '32'; // Half the hash.
+const prefixChecker = new RegExp('^[\\dA-F]{' + minimumPrefix + ',' + maximumPrefix + '}$', 'i');
```

Comment:
Regex checks for upper case, but prefix is converted to lowercase before being sent to getSkipSegmentsByHash - is there any point enforcing case when submitting a hash? If so i'd personally side on submitting in the same form as it's stored in the DB.